pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java

```diff
@@ -167,25 +168,47 @@ default T decode(byte[] bytes, byte[] schemaVersion) {
     }
 
     /**
-     * Create a Avro schema type by extracting the fields of the specified class.
+     * Create a  Avro schema type by default configuration of the class
      *
      * @param clazz the POJO class to be used to extract the Avro schema
      * @return a Schema instance
      */
     static <T> Schema<T> AVRO(Class<T> clazz) {
-        return DefaultImplementation.newAvroSchema(clazz);
+        return DefaultImplementation.newAvroSchema(new SchemaDefinition<>(clazz));
     }
 
     /**
-     * Create a JSON schema type by extracting the fields of the specified class.
+     * Create a Avro schema type with schema definition
+     *
+     * @param schemaDefinition the definition of the schema
+     * @return a Schema instance
+     */
+    static <T> Schema<T> AVRO(SchemaDefinition<T> schemaDefinition) {
```

Comment:
Instead of having the `SchemaDefinition<T>`, what about leaving the pojo and adding the options like: 

```java
Schema.AVRO(MyObject.class, AvroDefinition.builder().allowNull(false).build());
```
This could be extended in #3766  with: 

```java
Schema.AVRO(MyObject.class, AvroDefinition.fromSchema("...json definition..."));
```

Reply:
I don't think `AvroDefinition` is a good name. it is confusing - does it mean it is for avro schema, or it is for generated the avro formatted schema. I think `SchemaDefinition` is much better.

If we agree on `SchemaDefinition`, pojo class is better to be part of it. that means the schema definition is generated from pojo. if pojo class is omitted, the schema definition is generated from the json string.

Reply:
> I don't think AvroDefinition is a good name. it is confusing - does it mean it is for avro schema, or it is for generated the avro formatted schema. I think SchemaDefinition is much better.

Sure that's probably not a good naming. Though I'd be careful in using constructor approach since it makes it harder to extend, compared to using builder pattern.

> if pojo class is omitted, the schema definition is generated from the json string. 

Ok that would make the `Producer<Object>` which should still be fine (in the Avro case) since the avro schema is the one that is really deciding the serialization.

Reply:
Okay If I understand this correctly, the comment is changing the SchemaDefinition to use a builder pattern. correct?